message_test.go

```diff
@@ -53,3 +58,388 @@ func TestMessageThreadID(t *testing.T) {
+	}
+	thread := &Thread{}
+	if !threads.Next(thread) {
+		t.Fatalf("threads.Next(thread): unable to fetch the first and only thread")
+	}
+	msgs := thread.Messages()
+	msg := &Message{}
+	if !msgs.Next(msg) {
+		t.Fatalf("msgs.Next(msg): unable to fetch the first message in the thread")
+	}
+	if want, got := time.Unix(1258500098, 0), msg.Date(); want.Unix() != got.Unix() {
+		t.Errorf("msg.Date(): want %s got %s", want, got)
+	}
+}
+
+func TestMessageHeader(t *testing.T) {
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	qs := "id:1258471718-6781-2-git-send-email-dottedmag@dottedmag.net"
+	threads, err := db.NewQuery(qs).Threads()
+	if err != nil {
+		t.Fatalf("error getting the threads: %s", err)
+	}
+	thread := &Thread{}
+	if !threads.Next(thread) {
+		t.Fatalf("threads.Next(thread): unable to fetch the first and only thread")
+	}
+	msgs := thread.Messages()
+	msg := &Message{}
+	for msgs.Next(msg) {
+		if msg.ID() == "1258471718-6781-2-git-send-email-dottedmag@dottedmag.net" {
+			break
+		}
+	}
+	for _, hn := range []string{"References", "references"} {
+		if want, got := "<1258471718-6781-1-git-send-email-dottedmag@dottedmag.net>", msg.Header(hn); want != got {
+			t.Errorf("msg.Header(%q): want %s got %s", hn, want, got)
+		}
+	}
+}
+
+func TestMessageTags(t *testing.T) {
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	qs := "id:1258471718-6781-2-git-send-email-dottedmag@dottedmag.net"
+	threads, err := db.NewQuery(qs).Threads()
+	if err != nil {
+		t.Fatalf("error getting the threads: %s", err)
+	}
+	thread := &Thread{}
+	if !threads.Next(thread) {
+		t.Fatalf("threads.Next(thread): unable to fetch the first and only thread")
+	}
+	msgs := thread.Messages()
+	msg := &Message{}
+	for msgs.Next(msg) {
+		if msg.ID() == "1258471718-6781-2-git-send-email-dottedmag@dottedmag.net" {
+			break
+		}
+	}
+
+	ts := msg.Tags()
+	tag := &Tag{}
+	var tags []string
+	for ts.Next(tag) {
+		tags = append(tags, tag.Value)
+	}
+	if want, got := []string{"inbox", "unread"}, tags; !reflect.DeepEqual(want, got) {
+		t.Errorf("msg.Tags(): want %v got %v", want, got)
+	}
+}
+
+func TestMessageAddRemoveTagReadonlyDB(t *testing.T) {
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	qs := "id:1258471718-6781-2-git-send-email-dottedmag@dottedmag.net"
+	threads, err := db.NewQuery(qs).Threads()
+	if err != nil {
+		t.Fatalf("error getting the threads: %s", err)
+	}
+	thread := &Thread{}
+	if !threads.Next(thread) {
+		t.Fatalf("threads.Next(thread): unable to fetch the first and only thread")
+	}
+	msgs := thread.Messages()
+	msg := &Message{}
+	for msgs.Next(msg) {
+		if msg.ID() == "1258471718-6781-2-git-send-email-dottedmag@dottedmag.net" {
+			break
+		}
+	}
+
+	ts := msg.Tags()
+	tag := &Tag{}
+	var tags []string
+	for ts.Next(tag) {
+		tags = append(tags, tag.Value)
+	}
```

Comment:
Okay, I'm sold on it being in this package rather than an add-on. I think you're right though, we'll do it in a separate pr. I'm going to open an issue so it doesn't get forgotten.